### PR TITLE
Add Preflight Checks to Graylog startup

### LINKED ISSUE #5248: Create pre-flight checks for Elasticsearch and MongoDB versions
Graylog is currently not doing early pre-flight checks for required versions of Elasticsearch and MongoDB.

We only rely on driver compatibility checks and some manual checks in some parts of our code base at the moment. This might cause problems when users are upgrading to a new Graylog release without upgrading required services like Elasticsearch or MongoDB.
Especially using the wrong Elasticsearch version can break Graylog setups including data loss.

To avoid any problems, Graylog should run very early pre-flight checks for required services. These checks should run before we start any internal services, probably even before we create the dependency injector instance.
If one of these pre-flight checks fails, the Graylog process should be stopped and the user be notified about the problems.

## Proposed Checks & Tasks

- Check Elasticsearch version
- Check MongoDB version
- Check if there is enough space for the message journal and maybe check that the journal is not on a network filesystem like NFS (if that's possible)
- Run database migrations (pure MongoDB client ones) (see https://github.com/Graylog2/graylog2-server/issues/6671)
- Run ES index mapping migrations

### PR DESCRIPTION
**Pull request by mpfz0r**
### Server Preflight Checks
When Graylog is starting up, perform connectivity and version
checks for MongoDB and {Elastic,Open}Search.

By default Graylog will wait indefinitely until it can reach MongoDB and ES.
Once the connections are established we assert the versions
for compatibility. If those fail, we abort the startup without
running any further actions, e.g. migrations, periodicals, etc.
This change introduces two configuration flags:

 - `skip_preflight_checks`  which is false by default
 - `mongodb_version_probe_attempts` which is 0 (indefinitely) by default.

If an on-disk Kafka Journal is used, the preflight check will also assert that the
`message_journal_dir` is writable and has enough space to contain `message_journal_max_size`.
Also it should not be hosted on a network filesystem: (see https://github.com/oshi/oshi/blob/oshi-parent-5.3.7/oshi-core/src/main/java/oshi/software/os/linux/LinuxFileSystem.java#L159-L160)

### Pluggable Prefllight Checks
The Plugin API has been extended to allow for Plugins to register their own preflight checks:
To do this, Plugins need to implement `PreflightCheckModule` and return it via `Plugin#preflightCheckmodules()` :
The Module sets up a minimal Guice Binding that will be used to create an injector to create the instances needed to perform preflight checks.
 
### Other changes
Gracefully handle `null` Messages in `RawMessage#decode()`

Fixes #5248

### CODE REVIEW

**Review by thll — CHANGES_REQUESTED** [review]
Works well 👍 

I have some concerns about the MongoDB network file system check. The other remarks are just nitpicks.

File: graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/ServerPreflightCheck.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+                        journalMaxSize.toMegabytes()
+                ));
+            }
+            if (journalFs.typeName() != null && journalFs.typeName().equals("Network Disk")) {
+                throw new PreflightCheckException(StringUtils.f(
+                        "Journal directory <%s> should not be on a network file system (%s)!",
+                        journalDirectory.toAbsolutePath(),
+                        journalFs.sysTypeName()
+                ));
+            }
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/ServerPreflightCheck.java:182]: Once I have the "fresh installation detection" in place, we could fail only on new setups...

Any opinion from @bernd on this?

File: graylog2-server/src/main/java/org/graylog2/bootstrap/ServerBootstrap.java [hunk 2]
```diff
@@ -115,16 +121,37 @@ protected void beforeStart(TLSProtocolsConfiguration tlsProtocolsConfiguration,
...
+                new ServerStatusBindings(capabilities()),
+                new ConfigurationModule(configuration),
+                new SystemStatsModule(configuration.isDisableNativeSystemStatsCollector()),
+                new ServerPreflightChecksModule(configuration),
+                new Module() {
+                    @Override
+                    public void configure(Binder binder) {
+                        preflightCheckModules.forEach(binder::install);
+                    }
+                });
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/bootstrap/ServerBootstrap.java:153]: I left the non-lambda version intentionally, because it's easier to understand, IMO.
But I'm fine either way.

File: graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/ServerPreflightCheck.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/ServerPreflightCheck.java:182]: @mpfz0r @thll Good point. Let's enable that check only for new installations. :+1:

**Review by thll — CHANGES_REQUESTED** [review]

File: graylog2-server/src/main/java/org/graylog2/storage/versionprobe/VersionProbe.java [hunk 3]
```diff
@@ -77,7 +77,8 @@ public VersionProbe(ObjectMapper objectMapper,
                     .withRetryListener(new RetryListener() {
                         @Override
                         public <V> void onRetry(Attempt<V> attempt) {
-                            if (attempt.getAttemptNumber() == 1) {
+                            final V result = attempt.getResult();
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/storage/versionprobe/VersionProbe.java]: good catch :+1:

File: graylog2-server/src/main/java/org/graylog2/bootstrap/ServerBootstrap.java [hunk 4]
```diff
@@ -115,16 +121,37 @@ protected void beforeStart(TLSProtocolsConfiguration tlsProtocolsConfiguration,
...
+                .flatMap(Collection::stream).collect(Collectors.toList());
+
+        final Injector injector = getPreflightInjector(preflightCheckModules);
+
+        // Run server preflight checks
+        injector.getInstance(ServerPreflightCheck.class).runChecks(configuration);
+
+        // Run preflight checks from plugins
+        preflightCheckModules.forEach(m -> m.doCheck(injector));
+    }
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/bootstrap/ServerBootstrap.java:139]: There are currently two ways of running pre-flight checks, the built-in server checks and checks from plugins.

How about introducing some "pre-flight check service" that gets all checks injected, including the server ones?

That way, we can split up the `ServerPreflightCheck` class into smaller modules that we register the same way plugins would do.

File: graylog2-server/src/main/java/org/graylog2/plugin/PreflightCheckModule.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ *     Set up a minimal Guice Binding that will be used to create an injector
+ *     to create the instances needed to perform preflight checks.
+ * </ul>
+ * <ul>
+ *     Run the actual preflight checks by implementing {@link PreflightCheckModule#doCheck(Injector)}
+ *     The Injector can be used to create the necessary test instances.
+ * </ul>
+ */
+public abstract class PreflightCheckModule extends PluginModule {
+    public abstract void doCheck(Injector injector) throws PreflightCheckException;
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/plugin/PreflightCheckModule.java:37]: I think having the Guice module execute the checks is mixing up responsibilities. If we should decide to go with my suggestion above, having a service that runs all checks, we don't need to add a particular pre-flight check module that executes the checks. We would only need to create a pre-flight check module that registers the checks into a set or map binder.

File: graylog2-server/src/main/java/org/graylog2/plugin/journal/RawMessage.java [hunk 6]
```diff
@@ -129,6 +129,10 @@ public RawMessage(JournalMessage journalMessage, Object messageQueueId) {
 
     @Nullable
     public static RawMessage decode(final byte[] buffer, final Object messageQueueId) {
+        if (buffer == null) {
+            log.error("Cannot read <null> message from journal, ignoring this message.");
+            return null;
+        }
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/plugin/journal/RawMessage.java:135]: Is this change related to the PR?

File: graylog2-server/src/main/java/org/graylog2/database/MongoDBVersionCheck.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import org.graylog2.bootstrap.preflight.PreflightCheckException;
+import org.graylog2.shared.utilities.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+
+public class MongoDBVersionCheck {
+    private static final Logger LOG = LoggerFactory.getLogger(MongoDBVersionCheck.class);
+    private static final Version MINIMUM_MONGODB_VERSION = Version.forIntegers(2, 4);
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/database/MongoDBVersionCheck.java:35]: The minimum MongoDB version is 3.6 or even 4.0. Our documentation says 3.6.

**Review by bernd — CHANGES_REQUESTED** [review]
Thank you! :pray: A few comments.

**Review by thll — APPROVED** [review]

File: graylog2-server/src/main/java/org/graylog2/database/MongoDBVersionCheck.java (hunk 7 above)
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/database/MongoDBVersionCheck.java:35]: I took the check from existing code. But sure, I can adopt it.

File: graylog2-server/src/main/java/org/graylog2/plugin/journal/RawMessage.java (hunk 6 above)
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/plugin/journal/RawMessage.java:135]: Kind of, the related cloud plugin one uses a null Message as a Kafka canary.
We might want to change the canary.
But avoiding a NPE here is probably a good idea anyway..

File: graylog2-server/src/main/java/org/graylog2/shared/messageq/MessageQueueWriter.java [hunk 8]
```diff
@@ -30,6 +31,9 @@
 
     void write(List<RawMessageEvent> entries) throws MessageQueueException;
 
+    default void preFlightCheck() throws PreflightCheckException {
+    }
+
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/shared/messageq/MessageQueueWriter.java:36]: did you look at the cloud-plugin pr?

[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog2/shared/messageq/MessageQueueWriter.java:36]: Yep, I know that the Kafka message queue writer uses this. :slightly_smiling_face: I still think it would be nicer to decouple the pre-flight check logic from the `MessageQueueWriter` interface.

File: graylog2-server/src/main/java/org/graylog2/bootstrap/ServerBootstrap.java (hunk 4 above)
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/bootstrap/ServerBootstrap.java]: Great idea :+1: Done :heavy_check_mark:

File: graylog2-server/src/main/java/org/graylog2/shared/messageq/MessageQueueWriter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/shared/messageq/MessageQueueWriter.java]: Ok, done using a cast.

**Review by bernd — APPROVED** [review]
LGTM :+1: Thank you! :pray:
